src/databricks/labs/ucx/source_code/linters/directfs.py

```diff
@@ -72,43 +72,36 @@ class _DetectDirectFsAccessVisitor(TreeVisitor):
     def __init__(self, session_state: CurrentSessionState, prevent_spark_duplicates: bool) -> None:
         self._session_state = session_state
         self._directfs_nodes: list[DirectFsAccessNode] = []
-        self._reported_locations: set[tuple[int, int]] = set()
         self._prevent_spark_duplicates = prevent_spark_duplicates
 
     def visit_call(self, node: Call):
         for arg in node.args:
-            self._visit_arg(arg)
+            self._visit_arg(node, arg)
 
-    def _visit_arg(self, arg: NodeNG):
+    def _visit_arg(self, call: Call, arg: NodeNG):
         try:
             for inferred in InferredValue.infer_from_node(arg, self._session_state):
                 if not inferred.is_inferred():
                     logger.debug(f"Could not infer value of {arg.as_string()}")
                     continue
-                self._check_str_constant(arg, inferred)
+                self._check_str_arg(call, arg, inferred)
         except InferenceError as e:
             logger.debug(f"Could not infer value of {arg.as_string()}", exc_info=e)
 
-    def visit_const(self, node: Const):
-        # Constant strings yield Advisories
-        if isinstance(node.value, str):
-            self._check_str_constant(node, InferredValue([node]))
-
-    def _check_str_constant(self, source_node: NodeNG, inferred: InferredValue):
-        if self._already_reported(source_node, inferred):
-            return
-        # don't report on JoinedStr fragments
-        if isinstance(source_node.parent, JoinedStr):
-            return
+    def _check_str_arg(self, call_node: Call, arg_node: NodeNG, inferred: InferredValue):
         value = inferred.as_string()
         for pattern in DIRECT_FS_ACCESS_PATTERNS:
             if not pattern.matches(value):
                 continue
-            # avoid false positives with relative URLs
-            if self._is_http_call_parameter(source_node):
+            # don't capture calls not originating from spark or dbutils
+            is_from_spark = False
+            is_from_db_utils = Tree(call_node).is_from_module("dbutils")
+            if not is_from_db_utils:
+                is_from_spark = Tree(call_node).is_from_module("spark")
```

Comment:
Why can `is_from_spark` only be `True` when `is_from_db_utils` is `False`. I would expect those variables to be independent. If that it is the case, this dependency introduced on line 99 creates a bug on line 104

Reply:
They are independent, but if either is False we never reach line 104

Reply:
Do you mean:

> They are independent, but if **both are** False we never reach line 104

That is what I read: the "and" in the code with negations on both sides

Reply:
Yup sorry

Reply:
The purpose here is to minimize calls to `is_from_module` because they're not cheap

Reply:
> The purpose here is to minimize calls to is_from_module because they're not cheap

Could you add that as a comment? It is fine with me, just want to be sure the logic is correct though

Reply:
Done